axum-extra/src/response/attachment.rs

```diff
@@ -0,0 +1,119 @@
+use axum::response::IntoResponse;
+use http::{header, HeaderMap, HeaderValue};
+use tracing::trace;
+
+/// A file attachment response type
+///
+/// This type will set the `Content-Disposition` header for this response. In response a webbrowser
+/// will download the contents locally.
+///
+/// Use the `filename` and `content_type` methods to set the filename or content-type of the
+/// attachment. If these values are not set they will not be sent.
+///
+///
+/// # Example
+///
+/// ```rust
+///  use axum::{http::StatusCode, routing::get, Router};
+///  use axum_extra::response::Attachment;
+///
+///  async fn cargo_toml() -> Result<Attachment<String>, (StatusCode, String)> {
+///      let file_contents = tokio::fs::read_to_string("Cargo.toml")
+///          .await
+///          .map_err(|err| (StatusCode::NOT_FOUND, format!("File not found: {err}")))?;
+///      Ok(Attachment::new(file_contents)
+///          .filename("Cargo.toml")
+///          .content_type("text/x-toml"))
+///  }
+///
+///  let app = Router::new().route("/Cargo.toml", get(cargo_toml));
+///  let _: Router = app;
+/// ```
+///
+/// Hyper will set the `Content-Length` header if it knows the length. To manually set this header
+/// use the [`Attachment`] type in a tuple.
+///
+/// # Note
+/// If the content length is known and this header is manually set to a different length hyper
+/// panics.
+///
+/// ```rust
+/// async fn with_content_length() -> impl IntoResponse {
+///     (
+///         [(header::CONTENT_LENGTH, 3)],
+///         Attachment::new([0, 0, 0])
+///             .filename("Cargo.toml")
+///             .content_type("text/x-toml"),
+///     )
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Attachment<T> {
+    inner: T,
+    filename: Option<HeaderValue>,
+    content_type: Option<HeaderValue>,
+}
+
+impl<T: IntoResponse> Attachment<T> {
+    /// Creates a new [`Attachment`].
+    pub fn new(inner: T) -> Self {
+        Self {
+            inner,
+            filename: None,
+            content_type: None,
+        }
+    }
+
+    /// Sets the filename of the [`Attachment`]
+    ///
+    /// This updates the `Content-Disposition` header to add a filename.
+    pub fn filename<H: TryInto<HeaderValue>>(mut self, value: H) -> Self {
+        if let Some(filename) = value.try_into().ok() {
+            self.filename = Some(filename);
+        } else {
+            trace!("Attachment filename contains invalid characters");
+        }
```

Comment:
If calling this fn multiple times, I think it makes more sense if the last call fails, that the field is `None`. Wdyt?

```suggestion
        self.filename = if let Ok(filename) = value.try_into() {
            Some(filename)
        } else {
            trace!("Attachment filename contains invalid characters");
            None
        };
```

Reply:
Yes, that seems fair. That makes this function more predictable.